### PR TITLE
Fix stale MCP session diagnostics and reconnect log spam

### PR DESCRIPTION
**Pull request by dialupdisaster**
## Summary
- Adds a structured, machine-actionable stale streamable-HTTP MCP session error so clients/LLMs can reinitialize instead of repeatedly retrying an invalid `mcp-session-id`.
- Reduces editor WebSocket reconnect log spam by extending capped backoff to 60s and logging only initial/periodic reconnect attempts while preserving indefinite automatic reconnect.
- Adds unit and GDScript coverage for stale-session response rewriting and reconnect backoff/log-throttling behavior.

## Problem
When a streamable-HTTP MCP client keeps using an `mcp-session-id` that the MCP SDK no longer knows about, the Python MCP SDK rejects the request before Godot AI tool handlers run. The current response is a plain JSON-RPC error:

```json
{
  "code": -32600,
  "message": "Session not found"
}
```

That error is technically correct, but it gives clients and LLM agents no clear recovery action. In practice, a client can keep retrying the same stale session forever even though a fresh MCP session would work.

Separately, the editor WebSocket reconnect loop currently retries forever with a 10-second cap and logs every attempt. If the WebSocket cannot reconnect overnight, this can produce thousands of repeated reconnect log lines.

## Why not silently recover the old MCP session?
Server-side resurrection of a stale streamable-HTTP session is not safe here. The missing `mcp-session-id` refers to transport/session-manager state that no longer exists. Reusing or silently recreating that session ID risks violating MCP initialization/session semantics.

The correct recovery is for the MCP client to initialize a fresh streamable-HTTP MCP session. This PR makes that explicit and machine-readable.

## New stale-session response
For the MCP SDK's stale session response, Godot AI now preserves the HTTP 404 and JSON-RPC `-32600` error code, but rewrites the message and adds structured recovery data:

```json
{
  "jsonrpc": "2.0",
  "id": "server-error",
  "error": {
    "code": -32600,
    "message": "MCP session expired or was not found; reinitialize the streamable HTTP session",
    "data": {
      "recoverable": true,
      "action": "reinitialize_mcp_session",
      "reason": "stale_streamable_http_session"
    }
  }
}
```

Clients and LLM tooling can key off:

```text
error.data.action == "reinitialize_mcp_session"
```

and create a fresh MCP session instead of retrying the same stale one.

## Reconnect behavior change
The editor WebSocket reconnect loop still retries indefinitely, but less noisily:

- Backoff changes from `[1, 2, 4, 8, 10]` to `[1, 2, 4, 8, 16, 30, 60]`.
- Attempts 1–5 log normally for immediate diagnostics.
- Later attempts only log every 10th attempt.
- Reconnect still recreates a fresh `WebSocketPeer` before each attempt; the code now documents why.

This keeps recovery behavior while avoiding runaway log spam.

## Implementation notes
- `StaleMcpSessionDiagnosticMiddleware` wraps HTTP / streamable-HTTP ASGI apps and only rewrites the SDK's exact stale-session JSON-RPC response.
- Non-stale 404 responses, non-404 responses, and streaming/SSE bodies are preserved.
- `GodotAIFastMCP` applies the middleware to direct `server.run(transport="streamable-http")` and reload ASGI paths.

## Validation
- `uv run pytest tests/unit/test_asgi_session_diagnostics.py tests/unit/test_cli_reload.py` — 16 passed
- `uv run ruff check src tests` — passed
- `uv run pytest tests` — 624 passed

## Caveat
This does not make stale streamable-HTTP sessions resumable after server/plugin restart. It gives clients the correct recovery signal. Clients that ignore JSON-RPC `error.data` may still need a manual MCP reconnect/reinitialize.

### CODE REVIEW

**Comment by dialupdisaster** [comment]
Quick update: I pushed two follow-up commits after CI/Codecov feedback:

- `f1c963a` fixes the Godot/editor CI startup failure by making `StaleMcpSessionDiagnosticMiddleware` transparent to FastMCP app attributes like `.state`. This preserves the wrapper behavior while allowing FastMCP's HTTP runner to start normally.
- `db29be6` addresses the Codecov patch coverage issue with useful middleware edge-case tests rather than filler: chunked stale-session responses, non-HTTP passthrough, unknown ASGI message passthrough, non-session JSON-RPC 404 passthrough, and SSE non-wrapping behavior. I also removed a couple of unreachable defensive branches in the private helper path.

Local validation after the coverage update:

- `uv run ruff check src tests` passed
- focused session/CLI tests: `22 passed`
- targeted coverage shows `src/godot_ai/asgi.py` at 100%
- full Python suite: `630 passed`

The new workflow checks are still present as requested, but upstream Actions may still need approval to run on the fork PR.

**Comment by dsarno** [comment]
Hey @dialupdisaster — I reviewed this PR and Copilot's feedback. I agree with all four of Copilot's comments and put together a patch you can apply on top of `db29be6` so credit stays on this PR. I'll close my parallel PR (#227) once you've grabbed this.

## What this addresses

1. **`_send_response` typed `start_message` as `Message | None` but always uses it.** Tightened the param type to non-`Optional` (the call site already guarantees non-None via the `if start_message is None` guard in `capture_send`).
2. **`_is_sdk_session_not_found` was too broad** — matched on `error.message` alone. Now also requires `id == "server-error"` and `error.code == -32600`. An unrelated JSON-RPC 404 that happens to share the message text won't be rewritten.
3. **Workflow duplication.** `pr-checks.yml` overlaps with `ci.yml` (which already runs lint + the full pytest suite on PRs across two Python versions with coverage). Deleting it removes the duplicated CI runtime.
4. **Middleware buffered all HTTP 404s, not just stale-session ones.** Gated buffering on the request having an `mcp-session-id` header — the SDK only emits "Session not found" when the client sent that header, so any 404 without one passes through with original streaming/timing.

## Test changes

- Added `test_stale_mcp_session_diagnostic_passes_through_when_request_has_no_session_id` — proves the gate doesn't rewrite even a perfect stale-looking body when the header is absent.
- Added `test_stale_mcp_session_diagnostic_requires_matching_jsonrpc_id` — proves the tighter matcher rejects a payload whose only difference from the SDK's response is the `id` field.
- Added `test_stale_mcp_session_diagnostic_ignores_non_dict_jsonrpc_payloads` — covers the matcher's "not a dict" and "error is a string" rejection branches.
- Added the `mcp-session-id` header to the chunked-body, 200-OK, SSE, extension-message, plain-text 404, and other-JSON-RPC-404 tests so they actually exercise the rewriter's reject path (not just the gate).

## Validation

- `ruff check src/ tests/` — clean
- `pytest -q` — 633 passed
- `src/godot_ai/asgi.py` patch coverage: 100% (105/105 stmts)

## Not touched

- `connection.gd` reconnect changes — Copilot didn't flag them and they look right.
- The `__getattr__` proxy on the middleware — covered by `test_stale_mcp_session_diagnostic_preserves_fastmcp_app_state` and required for FastMCP's runner to read `app.state`.

## Patch

Apply with `git am` (it's a `git diff` so use `git apply` instead, or save as a file and `git apply`):

```diff
diff --git a/.github/workflows/pr-checks.yml b/.github/workflows/pr-checks.yml
deleted file mode 100644
index 317be45..0000000
--- a/.github/workflows/pr-checks.yml
+++ /dev/null
@@ -1,39 +0,0 @@
-name: PR Checks
-
-on:
-  workflow_dispatch:
-  push:
-    branches: [main]
-  pull_request:
-    branches: [main]
-
-jobs:
-  python-checks:
-    name: Python checks
-    runs-on: ubuntu-latest
-
-    steps:
-      - name: Check out repository
-        uses: actions/checkout@v6
-
-      - name: Set up Python
-        uses: actions/setup-python@v6
-        with:
-          python-version: "3.13"
-
-      - name: Set up uv
-        uses: astral-sh/setup-uv@v7
-        with:
-          enable-cache: true
-
-      - name: Install dependencies
-        run: uv sync --extra dev
-
-      - name: Lint
-        run: uv run ruff check src tests
-
-      - name: Run focused session recovery tests
-        run: uv run pytest tests/unit/test_asgi_session_diagnostics.py tests/unit/test_cli_reload.py
-
-      - name: Run Python test suite
-        run: uv run pytest tests
diff --git a/src/godot_ai/asgi.py b/src/godot_ai/asgi.py
index 3c147ae..e089afd 100644
--- a/src/godot_ai/asgi.py
+++ b/src/godot_ai/asgi.py
@@ -52,7 +52,10 @@ class StaleMcpSessionDiagnosticMiddleware:
         return getattr(self.app, name)
 
     async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
-        if scope["type"] != "http":
+        # The SDK only emits "Session not found" when the client sent an
+        # mcp-session-id header. Skip buffering for everything else so unrelated
+        # HTTP 404s keep their original streaming/timing semantics.
+        if scope["type"] != "http" or not _request_has_mcp_session_id(scope):
             await self.app(scope, receive, send)
             return
 
@@ -81,11 +84,11 @@ class StaleMcpSessionDiagnosticMiddleware:
 
     async def _send_response(
         self,
-        start_message: Message | None,
+        start_message: Message,
         body: bytes,
         send: Send,
     ) -> None:
-        rewritten = self._rewrite_stale_session_body(start_message, body)
+        rewritten = self._rewrite_stale_session_body(body)
         response_body = rewritten if rewritten is not None else body
         headers = start_message.get("headers", [])
         if rewritten is not None:
@@ -95,7 +98,7 @@ class StaleMcpSessionDiagnosticMiddleware:
         await send(start_message)
         await send({"type": "http.response.body", "body": response_body, "more_body": False})
 
-    def _rewrite_stale_session_body(self, start_message: Message, body: bytes) -> bytes | None:
+    def _rewrite_stale_session_body(self, body: bytes) -> bytes | None:
         try:
             payload = json.loads(body.decode("utf-8"))
         except (UnicodeDecodeError, json.JSONDecodeError):
@@ -108,12 +111,16 @@ class StaleMcpSessionDiagnosticMiddleware:
         return json.dumps(payload, separators=(",", ":")).encode("utf-8")
 
     def _is_sdk_session_not_found(self, payload: Any) -> bool:
-        return (
-            isinstance(payload, dict)
-            and payload.get("jsonrpc") == "2.0"
-            and isinstance(payload.get("error"), dict)
-            and payload["error"].get("message") == "Session not found"
-        )
+        # Match on multiple stable SDK signals (code + id + message) to avoid
+        # rewriting unrelated JSON-RPC 404s that happen to share one field.
+        if not isinstance(payload, dict) or payload.get("jsonrpc") != "2.0":
+            return False
+        if payload.get("id") != "server-error":
+            return False
+        error = payload.get("error")
+        if not isinstance(error, dict):
+            return False
+        return error.get("code") == -32600 and error.get("message") == "Session not found"
 
     def _headers_without_content_length(self, headers: Any) -> list[tuple[bytes, bytes]]:
         return [
@@ -131,6 +138,10 @@ class StaleMcpSessionDiagnosticMiddleware:
         return [*headers, (b"content-type", b"application/json")]
 
 
+def _request_has_mcp_session_id(scope: Scope) -> bool:
+    return any(key.lower() == b"mcp-session-id" for key, _ in scope.get("headers", []))
+
+
 def _get_dev_transport() -> str:
     transport = os.environ.get(DEV_TRANSPORT_ENV, "streamable-http")
     if transport not in RELOADABLE_TRANSPORTS:
diff --git a/tests/unit/test_asgi_session_diagnostics.py b/tests/unit/test_asgi_session_diagnostics.py
index 5953607..0b858fe 100644
--- a/tests/unit/test_asgi_session_diagnostics.py
+++ b/tests/unit/test_asgi_session_diagnostics.py
@@ -110,7 +110,10 @@ async def test_stale_mcp_session_diagnostic_handles_chunked_stale_session_body()
 
     app = StaleMcpSessionDiagnosticMiddleware(sdk_stale_session_response)
 
-    sent = await _single_http_request(app)
+    sent = await _single_http_request(
+        app,
+        headers=[(b"mcp-session-id", b"stale-session-id")],
+    )
 
     assert sent[0]["status"] == 404
     assert (b"content-type", b"application/json") in sent[0]["headers"]
@@ -119,8 +122,45 @@ async def test_stale_mcp_session_diagnostic_handles_chunked_stale_session_body()
     assert body["error"]["data"]["action"] == "reinitialize_mcp_session"
 
 
+@pytest.mark.anyio
+async def test_stale_mcp_session_diagnostic_passes_through_when_request_has_no_session_id():
+    # Without an mcp-session-id header the SDK never emits its stale-session
+    # response, so the middleware must not buffer or rewrite anything.
+    async def stale_looking_response(scope, receive, send):
+        await send(
+            {
+                "type": "http.response.start",
+                "status": 404,
+                "headers": [(b"content-type", b"application/json")],
+            }
+        )
+        await send(
+            {
+                "type": "http.response.body",
+                "body": json.dumps(
+                    {
+                        "jsonrpc": "2.0",
+                        "id": "server-error",
+                        "error": {"code": -32600, "message": "Session not found"},
+                    }
+                ).encode(),
+                "more_body": False,
+            }
+        )
+
+    app = StaleMcpSessionDiagnosticMiddleware(stale_looking_response)
+
+    sent = await _single_http_request(app)
+
+    body = json.loads(sent[1]["body"])
+    assert body["error"] == {"code": -32600, "message": "Session not found"}
+    assert "data" not in body["error"]
+
+
 @pytest.mark.anyio
 async def test_stale_mcp_session_diagnostic_leaves_other_responses_unchanged():
+    # Use the session header so this exercises the buffering code path's
+    # non-404 short-circuit, not just the request-level gate.
     async def ok_response(scope, receive, send):
         await send(
             {
@@ -133,7 +173,10 @@ async def test_stale_mcp_session_diagnostic_leaves_other_responses_unchanged():
 
     app = StaleMcpSessionDiagnosticMiddleware(ok_response)
 
-    sent = await _single_http_request(app)
+    sent = await _single_http_request(
+        app,
+        headers=[(b"mcp-session-id", b"live-session-id")],
+    )
 
     assert sent == [
         {
@@ -160,7 +203,10 @@ async def test_stale_mcp_session_diagnostic_streams_non_404_responses_unchanged(
 
     app = StaleMcpSessionDiagnosticMiddleware(streaming_response)
 
-    sent = await _single_http_request(app)
+    sent = await _single_http_request(
+        app,
+        headers=[(b"mcp-session-id", b"live-session-id")],
+    )
 
     assert sent == [
         {
@@ -197,7 +243,10 @@ async def test_stale_mcp_session_diagnostic_passes_unhandled_asgi_messages_throu
 
     app = StaleMcpSessionDiagnosticMiddleware(extension_message_response)
 
-    sent = await _single_http_request(app)
+    sent = await _single_http_request(
+        app,
+        headers=[(b"mcp-session-id", b"live-session-id")],
+    )
 
     assert sent == [
         {
@@ -207,6 +256,59 @@ async def test_stale_mcp_session_diagnostic_passes_unhandled_asgi_messages_throu
     ]
 
 
+@pytest.mark.anyio
+async def test_stale_mcp_session_diagnostic_ignores_non_dict_jsonrpc_payloads():
+    # A 404 with a JSON body that isn't the SDK's stale-session shape (here a
+    # list, and a dict whose error field is a string) must pass through.
+    async def list_payload_response(scope, receive, send):
+        await send(
+            {
+                "type": "http.response.start",
+                "status": 404,
+                "headers": [(b"content-type", b"application/json")],
+            }
+        )
+        await send(
+            {
+                "type": "http.response.body",
+                "body": b'[{"jsonrpc":"2.0","error":{"message":"Session not found"}}]',
+                "more_body": False,
+            }
+        )
+
+    app = StaleMcpSessionDiagnosticMiddleware(list_payload_response)
+    sent = await _single_http_request(
+        app,
+        headers=[(b"mcp-session-id", b"stale-session-id")],
+    )
+    body = json.loads(sent[1]["body"])
+    assert isinstance(body, list)
+
+    async def string_error_response(scope, receive, send):
+        await send(
+            {
+                "type": "http.response.start",
+                "status": 404,
+                "headers": [(b"content-type", b"application/json")],
+            }
+        )
+        await send(
+            {
+                "type": "http.response.body",
+                "body": b'{"jsonrpc":"2.0","id":"server-error","error":"Session not found"}',
+                "more_body": False,
+            }
+        )
+
+    app = StaleMcpSessionDiagnosticMiddleware(string_error_response)
+    sent = await _single_http_request(
+        app,
+        headers=[(b"mcp-session-id", b"stale-session-id")],
+    )
+    body = json.loads(sent[1]["body"])
+    assert body["error"] == "Session not found"
+
+
 @pytest.mark.anyio
 async def test_stale_mcp_session_diagnostic_leaves_other_404_responses_unchanged():
     async def not_found_response(scope, receive, send):
@@ -221,7 +323,10 @@ async def test_stale_mcp_session_diagnostic_leaves_other_404_responses_unchanged
 
     app = StaleMcpSessionDiagnosticMiddleware(not_found_response)
 
-    sent = await _single_http_request(app)
+    sent = await _single_http_request(
+        app,
+        headers=[(b"mcp-session-id", b"stale-session-id")],
+    )
 
     assert sent == [
         {
@@ -254,13 +359,48 @@ async def test_stale_mcp_session_diagnostic_leaves_other_json_rpc_404_errors_unc
 
     app = StaleMcpSessionDiagnosticMiddleware(json_rpc_not_found_response)
 
-    sent = await _single_http_request(app)
+    sent = await _single_http_request(
+        app,
+        headers=[(b"mcp-session-id", b"stale-session-id")],
+    )
 
     body = json.loads(sent[1]["body"])
     assert body["error"] == {"code": -32000, "message": "Tool not found"}
     assert "data" not in body["error"]
 
 
+@pytest.mark.anyio
+async def test_stale_mcp_session_diagnostic_requires_matching_jsonrpc_id():
+    # An unrelated 404 that happens to share the SDK's "Session not found"
+    # message but uses a different JSON-RPC id must not be rewritten.
+    async def lookalike_response(scope, receive, send):
+        body = json.dumps(
+            {
+                "jsonrpc": "2.0",
+                "id": "request-1",
+                "error": {"code": -32600, "message": "Session not found"},
+            }
+        ).encode()
+        await send(
+            {
+                "type": "http.response.start",
+                "status": 404,
+                "headers": [(b"content-type", b"application/json")],
+            }
+        )
+        await send({"type": "http.response.body", "body": body, "more_body": False})
+
+    app = StaleMcpSessionDiagnosticMiddleware(lookalike_response)
+
+    sent = await _single_http_request(
+        app,
+        headers=[(b"mcp-session-id", b"stale-session-id")],
+    )
+
+    body = json.loads(sent[1]["body"])
+    assert "data" not in body["error"]
+
+
 def test_create_server_wraps_streamable_http_app_with_stale_session_diagnostic():
     server = create_server()
 
```

To apply:

```bash
# save the diff above as /tmp/copilot-fixes.patch, then:
git checkout fix/session-recovery-diagnostics
git apply /tmp/copilot-fixes.patch
git add -A && git commit -m "fix(asgi): address Copilot feedback on stale-session middleware"
git push
```

Or pull from `hi-godot/godot-ai:claude/review-pr-225-s250C` and squash my two commits onto your branch — your call.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by dsarno** [comment]
Nice work @dialupdisaster! Thanks for being the first contributor! 🚀🤘🏽🎉
